onnxruntime/core/session/inference_session_utils.cc

```diff
@@ -0,0 +1,281 @@
+      return Status::OK();
+
+    case 2:
+      LOGS(logger, INFO) << "Setting graph_optimization_level to ORT_ENABLE_EXTENDED";
+      session_options.graph_optimization_level = TransformerLevel::Level2;
+      return Status::OK();
+
+    case 3:
+      LOGS(logger, INFO) << "Setting graph_optimization_level to ORT_ENABLE_ALL";
+      session_options.graph_optimization_level = TransformerLevel::Level3;
+      return Status::OK();
+
+    default:
+      LOGS(logger, ERROR) << "Unsupported graph_optimization_level value in ORT config: " << value;
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Unsupported graph_optimization_level value in ORT config: ", value);
+  }
+}
+
+static Status set_enable_profiling(SessionOptions& session_options,
+                                   int value,
+                                   const logging::Logger& logger) {
+  switch (value) {
+    case 0:
+      LOGS(logger, INFO) << "Setting enable_profiling to false";
+      session_options.enable_profiling = false;
+      return Status::OK();
+
+    case 1:
+      LOGS(logger, INFO) << "Setting enable_profiling to true";
+      session_options.enable_profiling = true;
+      return Status::OK();
+
+    default:
+      LOGS(logger, ERROR) << "Unsupported value for enable_profiling option: " << value;
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported value for enable_profiling option: ", value);
+  }
+}
+
+//---------------------------------------------------
+//--- end of session options related helpers ---
+//---------------------------------------------------
+
+//---------------------
+//--- end of local helpers ---
+//---------------------
+
+static const std::string session_options_key = "session_options";
+
+Status parse_session_options_from_model_proto(const ONNX_NAMESPACE::ModelProto& model_proto,
+                                              SessionOptions& session_options,
+                                              const logging::Logger& logger) {
+  json json_obj;
+  bool session_options_found = false;
+
+  for (auto metadata_field : model_proto.metadata_props()) {
+    if (metadata_field.has_key() && metadata_field.key() == "ort_config") {
+      LOGS(logger, INFO)
+          << "Found session/run/environment configuration in the model file to be used while running the model";
+
+      auto status =
+          parse_json_and_search_for_key(metadata_field.value(), session_options_key, session_options_found, json_obj);
+
+      if (!status.IsOK()) {
+        LOGS(logger, ERROR) << "Could not parse session/run/environment configuration json in the model file";
+        return status;
+      }
+    }
+  }
+
+  if (!session_options_found) {
+    LOGS(logger, INFO) << "Did not find session options in the model file to be used while running the model";
+    return Status::OK();
+  }
+
+  auto& session_options_from_model = json_obj.at(session_options_key);
```

Comment:
const auto&. Check other places too.

Reply:
Fixed here and in other applicable places. Thanks.